### ISSUE TITLE
ERROR MESSAGE: Failed to query available provider packages in tf module. but works if written in the same file

### ISSUE THREAD

**Issue by corganfuzz** [Azure/terraform-provider-azapi#373]
### Describe the bug

` required_providers ` block doesnt seem to work at the module level. only in the same file level

TLDR can be found [here](https://stackoverflow.com/questions/73293980/terraform-failed-to-query-available-provider-packages-azapi)

this is the error message:

``` bash
Initializing provider plugins...
- Finding latest version of azure/azapi...
- Finding latest version of hashicorp/random...
- Finding latest version of hashicorp/azapi...
- Finding hashicorp/azurerm versions matching "3.73.0"...
- Installing azure/azapi v1.10.0...
- Installed azure/azapi v1.10.0 (signed by a HashiCorp partner, key ID 6F0B91BDE98478CF)
- Installing hashicorp/random v3.5.1...
- Installed hashicorp/random v3.5.1 (signed by HashiCorp)
- Installing hashicorp/azurerm v3.73.0...
- Installed hashicorp/azurerm v3.73.0 (signed by HashiCorp)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html
╷
│ Error: Failed to query available provider packages
│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/azapi: provider registry registry.terraform.io does not have a
│ provider named registry.terraform.io/hashicorp/azapi
```

### Steps to reproduce the behavior

1. First of all this is my folder strcuture:

```bash
├── env
│   ├── dev
│   │   ├── env-dev.tf
│   │   ├── locals.tf
│   │   └── providers.tf
├── main
│   ├── main.tf
│   └── variables.tf
├── modules
│   └── vm_scaleset
│       ├── main.tf
│       └── variables.tf
```

The error happens when I put my provider configuration in  `env/dev/providers.tf`

```hcl
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "=3.73.0"
    }
    azapi = {
      source  = "Azure/azapi"
      version = "~>1.10.0"
    }
  }
}

```

### WORKAROUND

The error disappears when I deleted azapi from `env/dev/providers.tf` and put it on `modules/vm_scaleset/providers.tf` or `modules/vm_scaleset/main.tf` for that matter

```hcl 
terraform {
  required_providers {
    azapi = {
      source = "Azure/azapi"
      version = "~>1.10.0"
    }
  }
}

# rest of the code ...
resource "azapi_resource_action" "ssh_public_key_gen" {
  type        = "Microsoft.Compute/sshPublicKeys@2023-07-01"
  resource_id = azapi_resource.ssh_public_key.id
  action      = "generateKeyPair"
  method      = "POST"
  response_export_values = ["publicKey"]
}
...
```

### EXPECTED BEHAVIOR:

Successful message but it should sahy this when I put the provider at the module level

```bash
Initializing provider plugins...
- Finding hashicorp/azurerm versions matching "3.73.0"...
- Finding latest version of azure/azapi...
- Finding latest version of hashicorp/random...
- Installing azure/azapi v1.10.0...
- Installed azure/azapi v1.10.0 (signed by a HashiCorp partner, key ID 6F0B91BDE98478CF)
- Installing hashicorp/random v3.5.1...
- Installed hashicorp/random v3.5.1 (signed by HashiCorp)
- Installing hashicorp/azurerm v3.73.0...
- Installed hashicorp/azurerm v3.73.0 (signed by HashiCorp)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

```

**Comment by ms-henglu** [comment]
Hello @corganfuzz ,

Thank you for taking time to report this issue!

From the error message, I saw a version of `azapi` is installed, but it couldn't find the `registry.terraform.io/hashicorp/azapi` which has an incorrect namespace "hashicorp", I guess there is a module that doesn't have the following provider definition:

```
terraform {
  required_providers {
    azapi = {
      source = "Azure/azapi"
      version = "~>1.10.0"
    }
  }
}
```


```shell
....

- Installing azure/azapi v1.10.0...
- Installed azure/azapi v1.10.0 (signed by a HashiCorp partner, key ID 6F0B91BDE98478CF)

....


│ 
│ Could not retrieve the list of available versions for provider
│ hashicorp/azapi: provider registry registry.terraform.io does not have a
│ provider named registry.terraform.io/hashicorp/azapi
```

**Comment by corganfuzz** [comment]
Yeah. It's like .. it's trying to combine hashicorp + azapi or something but only at the module level

**Comment by va-chopra** [comment]
hi any resolution for this?

**Comment by rpvsilva** [comment]
Hey there, probably you're using this provider inside a custom module, for that you need to initialize the provider inside that module too. Terraform handles providers as part of `hashicorp` namespace by default. That's why the output is showing that it is trying to get from `registry.terraform.io/hashicorp/azapi`. So, what you need to do is, on your custom module you need to add, again, the required provider:

```terraform
terraform {
  required_version = ">= 1.6.5"

  required_providers {
    azapi = {
      source = "Azure/azapi"
      version = "1.12.0"
    }
  }
}
```

**Comment by corganfuzz** [comment]
Yup. I did that. I also mentioned that on the WORKAROUND part in the beginning of this issue. This issue is to report a bug in which `azapi `should work exactly the same as `hashicorp/azurerm` and only be written once.

**Comment by ms-henglu** [comment]
Hi @corganfuzz , the reason is, unlike azurerm provider which is an official provider(whose namespace is `hashicorp`), the `azapi` provider is a partner provider and its namespace is `azure`. And the partner providers must specify their namespaces.

**Comment by ms-henglu** [comment]
I'll close this issue as it's resolved, but feel free to reopen it.

**Comment by bisskar** [comment]
This is not really helpful to add provider block in the modules, it returns errors about missing providers when you comment out a azapi resource.
